src/overlays/actors/ovl_En_River_Sound/z_en_river_sound.c

```diff
@@ -27,15 +30,208 @@ const ActorInit En_River_Sound_InitVars = {
 
+    f = sp_20.z * sp_2C.z + (sp_2C.x * sp_20.x + sp_2C.y * sp_20.y);
+    g = sp_14.z * sp_2C.z + (sp_2C.x * sp_14.x + sp_2C.y * sp_14.y);
+    h = sp_2C.z * sp_2C.z + (sp_2C.x * sp_2C.x + sp_2C.y * sp_2C.y);
+
+    if (g * f < 0) {
+        a3->x = sp_2C.x * (-f / h) + a0->x;
+        a3->y = sp_2C.y * (-f / h) + a0->y;
+        a3->z = sp_2C.z * (-f / h) + a0->z;
+
+        return 1;
+    } else {
+        return 0;
+    }
+}
+#else
 #pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_River_Sound/func_80AE6A54.s")
+#endif
+
+s32 func_80AE6BC0(Vec3s* points, s32 num_points, Vec3f* pos, Vec3f* res) {
+    s32 i;
+    s32 point_idx;
+    s32 sp_78[2] = { 0, 0 };
+    Vec3f point_loc;
+    Vec3f sp_60;
+    Vec3f sp_54;
+    Vec3f vec;
+    f32 point_dist = 10000.f;
+    Vec3s* point;
+
+    for (i = 0; i < num_points; i++) {
+        f32 d;
+        vec.x = points[i].x;
+        vec.y = points[i].y;
+        vec.z = points[i].z;
+        d = Math_Vec3f_DistXYZ(pos, &vec);
+        if (d < point_dist) {
+            point_dist = d;
+            point_idx = i;
+        }
+    }
+    if (point_dist >= 10000.f)
+        return 0;
+
+    point = &points[point_idx];
+    point_loc.x = point->x;
+    point_loc.y = point->y;
+    point_loc.z = point->z;
+    if (point_idx != 0) {
+        vec.x = point[-1].x;
+        vec.y = point[-1].y;
+        vec.z = point[-1].z;
+        sp_78[0] = func_80AE6A54(&vec, &point_loc, pos, &sp_54);
+    }
+    if (point_idx + 1 != num_points) {
+        vec.x = point[1].x;
+        vec.y = point[1].y;
+        vec.z = point[1].z;
+        sp_78[1] = func_80AE6A54(&point_loc, &vec, pos, &sp_60);
+    }
+
+    if (sp_78[0] && sp_78[1]) {
+        if (!func_80AE6A54(&sp_54, &sp_60, pos, res)) {
+            res->x = (sp_54.x + sp_60.x) * .5f;
+            res->y = (sp_54.y + sp_60.y) * .5f;
+            res->z = (sp_54.z + sp_60.z) * .5f;
+        }
+    } else if (sp_78[0]) {
+        res->x = sp_54.x;
+        res->y = sp_54.y;
+        res->z = sp_54.z;
+    } else if (sp_78[1]) {
+        res->x = sp_60.x;
+        res->y = sp_60.y;
+        res->z = sp_60.z;
+    } else {
+        res->x = point_loc.x;
+        res->y = point_loc.y;
+        res->z = point_loc.z;
+    }
+
+    return 1;
+}
+
+void EnRiverSound_Update(Actor* thisx, GlobalContext* globalCtx) {
+    Path* path;
+    Vec3f* pos;
+    Actor* player = &PLAYER->actor;
+    CollisionContext* colCtx;
+    s32 sp_34;
+
+    if (thisx->params == 0 || thisx->params == 4 || thisx->params == 5) {
+        path = &globalCtx->setupPathList[THIS->pathIndex];
+        pos = &thisx->world.pos;
+
+        if (func_80AE6BC0(SEGMENTED_TO_VIRTUAL(path->points), path->count, &player->world.pos, pos)) {
+            colCtx = &globalCtx->colCtx;
+
+            if (BgCheck_EntityRaycastFloor4(colCtx, &thisx->floorPoly, &sp_34, thisx, pos) != -32000.f)
+                THIS->b_14D = SurfaceType_GetConveyorSpeed(colCtx, thisx->floorPoly, sp_34);
```

Comment:
```suggestion
                this->b_14D = SurfaceType_GetConveyorSpeed(colCtx, thisx->floorPoly, sp_34);
```
And have `EnRiverSound* this = THIS;` at the top